quickjs/src/jniMain/kotlin/app/cash/quickjs/Zipline.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.quickjs
+
+import app.cash.quickjs.internal.bridge.InboundService
+import app.cash.quickjs.internal.bridge.OutboundClientFactory
+import java.util.concurrent.Executors
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.launch
+
+actual abstract class Zipline {
+  actual abstract val engineVersion: String
+
+  abstract val quickJs: QuickJs
+
+  abstract val dispatcher: CoroutineDispatcher
+
+  actual fun <T : Any> get(name: String, jsAdapter: JsAdapter): T {
+    error("unexpected call to Zipline.get: is the Zipline plugin configured?")
+  }
+
+  @PublishedApi
+  internal abstract fun <T : Any> get(
+    name: String,
+    outboundClientFactory: OutboundClientFactory<T>
+  ): T
+
+  actual fun <T : Any> set(name: String, jsAdapter: JsAdapter, instance: T) {
+    error("unexpected call to Zipline.set: is the Zipline plugin configured?")
+  }
+
+  @PublishedApi
+  internal abstract fun set(name: String, service: InboundService<*>)
+
+  companion object {
+    fun create(
+      dispatcher: CoroutineDispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher(),
+    ): Zipline {
+      // TODO(jwilson): avoid runBlocking if I’m already on our dispatcher??
+      return kotlinx.coroutines.runBlocking(dispatcher) {
+        val quickJs = QuickJs.create()
+        ZiplineJvm(quickJs, dispatcher)
+      }
+    }
+  }
+}
+
+/** Execute [block] on the JavaScript dispatcher and wait for it to return. */
+inline fun <T> Zipline.runBlocking(crossinline block: suspend Zipline.() -> T): T {
+  return kotlinx.coroutines.runBlocking(dispatcher) {
+    block()
+  }
+}
+
+/** Enqueue [block] to run on the JavaScript dispatcher and return immediately. */
+inline fun Zipline.launch(crossinline block: suspend Zipline.() -> Unit) {
+  CoroutineScope(EmptyCoroutineContext).launch(dispatcher) {
+    block()
+  }
+}
```

Comment:
I find both of these highly suspect and don't think either is something I want to expose in public API. Either the caller is in a coroutine scope in which they can `launch`/`async`, they're in a suspend function where they can `withContext`, or they're not in which case they can `runBlocking`.

These feel like helpers for our tests.

Reply:
Lemme move it to our testUtil. I think it probably goes away when I add the test dispatcher.